pkg/services/ngalert/state/historian/dashboard.go

```diff
@@ -0,0 +1,89 @@
+package historian
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	"github.com/patrickmn/go-cache"
+)
+
+const dashboardNotFound = int64(-1)
+
+const defaultDashboardCacheExpiry = 1 * time.Minute
+
+const minCleanupInterval = 1 * time.Second
+
+// dashboardResolver resolves dashboard UIDs to IDs with caching.
+type dashboardResolver struct {
+	dashboards dashboards.DashboardService
+	cache      *cache.Cache
+	lock       sync.Mutex
+	log        log.Logger
+}
+
+func newDashboardResolver(dbs dashboards.DashboardService, log log.Logger, expiry time.Duration) *dashboardResolver {
+	return &dashboardResolver{
+		dashboards: dbs,
+		cache:      cache.New(expiry, maxDuration(2*expiry, minCleanupInterval)),
+		log:        log,
+	}
+}
+
+// getId gets the ID of the dashboard with the given uid/orgID combination, or returns dashboardNotFound if the dashboard does not exist.
+func (r *dashboardResolver) getId(ctx context.Context, orgID int64, uid string) (int64, error) {
+	// Optimistically query without acquiring lock. This works because cache.Cache is thread-safe.
+	// That way, we don't block valid queries on unrelated cache misses.
+	// We use a double-checked lock to prevent queries from being needlessly ran.
+	if id, found := r.cache.Get(packCacheKey(orgID, uid)); found {
+		return id.(int64), nil
+	}
+
+	r.lock.Lock()
```

Comment:
Would it be better to use `golang.org/x/sync/singleflight` here? Unless I've misunderstood the code, this mutex prevents simultaneous queries of different dashboards. Is that intentional?

Reply:
This is an excellent suggestion, I didn't know this existed. I'll try it out.